File: interviews/A/VO/courseSchedule.py
from collections import deque
class Solution:
    """
    @param: numCourses: a total of n courses
    @param: prerequisites: a list of prerequisite pairs
    @return: true if can finish all courses or false
    """
    def canFinish(self, numCourses, prerequisites):
        # write your code here
        
        graph = self.build_graph(numCourses, prerequisites)
        
        return self.topo_sort(graph)
        
    def build_graph(self, numCourses, prerequisites):
        
        graph = {}
        
        for node in range(numCourses):
            
            if node not in graph:
            
                graph[node] = set()
            
        for post, pre in prerequisites:
            
            graph[pre].add(post)
            
        return graph
        
    def get_in_degrees(self, graph):
        
        in_degress = { node : 0 for node in graph}
        
        for node in graph:
            
            for neighbor in graph[node]:
                
                in_degress[neighbor] += 1 
                
        return in_degress
        
    def topo_sort(self, graph):
        
        in_degress = self.get_in_degrees(graph)
        
        queue = deque()
        
        for node in graph:
            
            if in_degress[node] == 0:
                
                queue.append(node)
                
             
        order = [] 
        
        while queue:
            
            node = queue.popleft()
            
            order.append(node)
            
            for neighbor in graph[node]:
                
                in_degress[neighbor] -= 1 
                
                if in_degress[neighbor] == 0:
                    
                    queue.append(neighbor)
                    
        if len(order) == len(graph):
            
            return True 
            
        return False 
            
        

File: interviews/Twitch/JavaTwitch/CourseSchedule.java
import java.util.*; 

public class CourseSchedule {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        
        HashMap<Integer, HashSet<Integer>> graph = this.buildGraph(prerequisites, numCourses); 
        
        return topo_sorting(graph);
        
    }
    
    private HashMap<Integer, HashSet<Integer>> buildGraph(int[][] pres, int num) {
        
        HashMap<Integer, HashSet<Integer>> graph = new HashMap<Integer, HashSet<Integer>>(); 
        
        for (int i = 0; i < num; i++) {
            
            graph.put(i, new HashSet<Integer>());
        }
        
        for (int[] pre : pres) {
            
            graph.get(pre[1]).add(pre[0]);
            
        }
        
        return graph;
    }
    
    private HashMap<Integer, Integer> getIndegrees(HashMap<Integer, HashSet<Integer>> graph) {
        
        HashMap<Integer, Integer> indegrees = new HashMap<>(); 
        
        for (Integer node : graph.keySet()) {
            
            indegrees.put(node, 0);
        }
        
        for (Integer node : graph.keySet()) {
            
            for (Integer neighbor : graph.get(node)) {
                
                indegrees.put(neighbor, indegrees.get(neighbor) + 1);
            }
        }
        
        return indegrees;
    }
    
    private boolean topo_sorting(HashMap<Integer, HashSet<Integer>> graph) {
        
        HashMap<Integer, Integer> indegrees = this.getIndegrees(graph);
        
        Queue<Integer> queue = new LinkedList<Integer>(); 
        
        for (Integer node : graph.keySet()) {
            
            if (indegrees.get(node) == 0) {
                
                queue.offer(node);
            }
        }
        
        List<Integer> order = new ArrayList<Integer>(); 
        
        while (!queue.isEmpty()) {
            
            int node = queue.poll(); 
            
            order.add(node); 
            
            for (int neighbor : graph.get(node)) {
                
                indegrees.put(neighbor, indegrees.get(neighbor) - 1);
                
                if (indegrees.get(neighbor) == 0) {
                    queue.offer(neighbor);
                }
            }
        }
        
        
        return order.size() == graph.size();
        
    }
}